### PR TITLE
CSS Variables

### PR DESCRIPTION
**Pull request by r3dDoX**

### CODE REVIEW

**Review by mohsenr — APPROVED** [review]
Nice post – short and to the point. I left some minor comments.

I didn’t leave an inline comment for this as I’d like to discuss it generally: Would be good to make sure we use the correct apostrophe character. That is “Let's” should be written as “Let’s”.

In our projects I advocate for unit tests that disallow ' and " in all copy. It wouldn’t be as easy here as we need to allow it in code but something to keep in mind as potential improvement later.

File: dark-mode-css-vars.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+---
+title: Moving from Sass to CSS variables to implement Dark Mode
+domain: campzulu.hashnode.dev
+tags: web-development, css, sass
+cover: https://cdn.hashnode.com/res/hashnode/image/unsplash/xcweYgakbRo/upload/v1660032971546/KNUGnflXq.jpeg?w=1600&h=840&fit=crop&crop=entropy&auto=compress,format&format=webp
+publishAs: r3dDoX
+ignorePost: true
+---
+
+# Why Dark Mode
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: Good point, changed that

File: dark-mode-css-vars.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+We are currently working on a machine that runs 24/7 at train stations all over Switzerland. The
+screen is fairly bright to help people navigate through our web application even when there is
+direct sunlight hitting the machine. For people with visual impairment, this brightness is a big
+issue since some have to go very close to the screen to properly read and interact with it. This is
+why we started developing a Dark Mode that is specifically designed for people with visual
+impairment.
+
+# Why CSS Vars
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: I agree, changed it to "Introducing CSS variables"

File: dark-mode-css-vars.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+We knew from the beginning, that implementing a Dark Mode will need a lot of trial and error due to
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: Changed it

File: dark-mode-css-vars.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+:root {
+    --test-var: #1e90ff;
+}
+
+.test {
+    width: var(--test-var);
+}
+``` 
+
+At a first glance the CSS syntax looks more bloated but with the good IDE support nowadays you don't
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md:62]: Good guy grammarly :)

File: dark-mode-css-vars.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    <div>
+        <p>This will have a papayawhip background</p>
+    </div>
+</main>
+```
+
+The paragraph that is wrapped in the `div` will get the overwritten `--bg-color` while the other
+paragraph gets the original value of the variable. Check it out on Codepen:
+
+%[https://codepen.io/r3dDoX/pen/WNzgYNL]
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md:100]: that is hashnode specific and should work on their platform

File: dark-mode-css-vars.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+### sass:math
+
+In Sass you can have a lot of calculations at build time which with CSS is not possible. All
+calculations have to be done with
+the `calc()` ([MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/calc)) function at runtime.
+
+### sass:color
+
+Sass has a lot of helper functions to adjust color values which currently is not possible with CSS.
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: Yeah, that sounds much better

File: dark-mode-css-vars.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+What we were mostly missing were functions like: `darken()` or `lighten()`. Right now we were able
+to work around this by having a few more variables holding these variables.
+
+In the future we should have this functionality (and much more) in CSS described very well in this
+blog post: [CSS relative colors](https://blog.jim-nielsen.com/2021/css-relative-colors/)
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: Absolutely

File: dark-mode-css-vars.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+:root[data-theme="dark"] {
+    --color-background: #000;
+    --color-text-primary: #fff;
+    /* ... */
+}
+```
+
+All our application has to do now is set the attribute `data-theme="dark"` on our root element when
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md]: Yes, totally

File: dark-mode-css-vars.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+are used correctly and work properly in the dark mode.
+
+# Conclusion
+
+CSS variables allowed us to incrementally introduce a dark mode while continuously testing with
+people that the contrasts on the machines work as expected. During our testing sessions we could
+just adapt certain variables at runtime allowing us to directly test their feedback.
+
+On top of that the tooling during build time as well as runtime in the browser is very compelling.
+Over all we only missed the ability of relative colors which should be arriving in CSS in the near
```
[the remark this answers is not in the document]
Comment (r3dDoX) [reply dark-mode-css-vars.md:151]: Oh I might have written that wrong for a while now...

**Comment by r3dDoX** [comment]
> Nice post – short and to the point. I left some minor comments.
> 
> I didn’t leave an inline comment for this as I’d like to discuss it generally: Would be good to make sure we use the correct apostrophe character. That is “Let's” should be written as “Let’s”.
> 
> In our projects I advocate for unit tests that disallow ' and " in all copy. It wouldn’t be as easy here as we need to allow it in code but something to keep in mind as potential improvement later.

Good question, I really didn't think about it at all. Mostly due to my laziness :) What do you think, is this going to be an issue if we just use the single quote?

**Comment by r3dDoX** [comment]
Alright, other than the single quote question I think I put all the feedback in 👍

File: dark-mode-css-vars.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+every component stylesheet at build time. Which means changing a color at runtime to test something
+would only work by changing it everywhere it is used.
+
+A more promising approach seemed to be CSS variables. At runtime we would still have the references
+instead of hardcoded values. The Dev Tools show these variables and you can modify them:
+
+![CSS Variables in Dev Tools](https://cdn.hashnode.com/res/hashnode/image/upload/v1660130533039/a3Y9ukL8u.png?auto=compress)
+
+This allows us to use the Dev Tools to change colors at runtime and test them on the real machines
+with real people. So let's look at how we can introduce CSS variables into our codebase.
```
Comment (mohsenr) [inline dark-mode-css-vars.md:32]:
```suggestion
with real people. So let’s look at how we can introduce CSS variables into our codebase.
```

File: dark-mode-css-vars.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+:root {
+    --test-var: #1e90ff;
+}
+
+.test {
+    width: var(--test-var);
+}
+``` 
+
+At a first glance, the CSS syntax looks more bloated but with the good IDE support nowadays you don't
```
Comment (mohsenr) [inline dark-mode-css-vars.md:62]:
```suggestion
At a first glance, the CSS syntax looks more bloated but with the good IDE support nowadays you don’t
```

**Review by mohsenr — COMMENTED** [review]
@r3dDoX yeah the suggestion of adding tests was definitely long term, don’t worry about it.

I’ve recommended a path to fix it for this post though. It’s just two characters 😊

**Pull request merged**